docs/extended-collection-types.md

```diff
@@ -0,0 +1,169 @@
+
+## Motivation
+
+The implementation of BYOND lists is highly inefficient in OpenDream as we can't just back it with a C# list or a C# dictionary. Furthermore, other collection types would offer performant convenience features for developers that they do not need to implement themselves, such as performant hashsets or stacks.
+
+## Design
+
+The current `list()` type will remain unchanged for BYOND compatibility, with the other types being added *in addition to* these legacy lists.
+
+The following collection types will essentially be exposed directly from C# to DM through a minimal wrapper API:
+
+- **Array**: A fixed-length collection of elements.
+- **Dictionary**: A collection of key-value pairs. Equivalent to existing associative lists, and will serve as the internal implementation of BYOND's planned `alist()` feature.
+- **HashSet**: A collection that contains no duplicate elements.
+- **ODList**: A dynamically resizable collection that is purely a list with none of the assoc-list functionality that BYOND lists have.
+- **Queue**: A first-in first-out (FIFO) collection.
+- **Stack**: A last-in first-out (LIFO) collection.
+
+Generally speaking, the helper procs from `list()` will also operate on other collections where applicable (e.g. `list.Join()` and `array.Join()` would be equivalent). A few collections will implement additional helpers as noted below.
+
+**Note:** Initializing OpenDream's collection types with a fixed size, for example `var/array/foo[10]`, will initialize it with a `capacity` of 10 but *will not* insert any `null` elements; the actual contents of the collection will be empty. This is **breaking behavior** from BYOND lists such as `var/list/L[10]`, which inserts 10 `null` elements into the list. BYOND lists in OpenDream will preserve this behavior for compatibility. Otherwise, see the "Collection Capacity API" subsection below for more information.
+
+### Array
+
+A fixed-length collection of elements. Almost identical to `list()` except it cannot be resized.
+
+Example usage:
+```
+var/array/foo = array("a","b","c")
+foo += "z" // Compile-time error, cannot resize arrays
+
+world.log << foo.Join() // Prints "abc"
+
+foo[1] = "z" // Valid modification
+```
+
+### Dictionary
+A collection of key-value pairs. Equivalent to existing associative lists, and will serve as the internal implementation of BYOND's planned `alist()` feature.
+
+Example usage:
+```
+var/dict/bar = array("a" = 1,"b" = 2,"c" = 3)
+bar += "d" // Compile-time error, no value specified
+
+bar["d"] = 4 // Valid modification
+
+world.log << bar["c"] // Prints 3
+
+world.log << bar["test"] // Runtime error, key not found
+
+bar.contains_key("test") // Returns FALSE
+
+bar.get_value("test") // Returns null, since the key doesn't exist
+
+world.log << bar[1] // Compile-time error, dicts can't be numerically indexed
```

Comment:
Why are we not allowing dictionaries to have a numerical key? That's how it works with the existing lists, but this is an annoying limitation with the current system and if there is going to be a specific dictionary type, I see no reason why we have to block of numerical keys (since we can now identify between what the user really wants based on whether or not they ask for a `dict` or an `array`)

Reply:
Dictionaries aren't ordered.

Reply:
I mean, what if we have cases where we want to actually have an numerical key. Within current byond, we have to cast the number to a string and store it in the dictionary.

An example code of something that I think would be really useful to have, that would be an improvement over current byond:
```
var/dict/player_locations = dict(4 = dict(8 = odlist(new /mob)), 9 = dict(4 = odlist(new /mob)))
// Find the player at 4, 8
var/odlist/players = player_locations[4][8]
```

Reply:
![image](https://github.com/OpenDreamProject/rfcs/assets/26465327/2d60b11b-0160-426b-82ae-646b6efcc70c)

Reply:
Even without typed collections I would expect a dictionary type to have arbitrarily-typed `DreamValue` keys rather than just strings. 

This would give similar reasons to use `dict` over `list` as there are to use JavaScript's `Map` type over normal JS objects.